rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+
+Optionally, we could accompany this with a simpler, low-abstraction alternative to stdenv, or an opt-in C compiler setup.
+
+It is worth noting that `stdenv.mkDerivation` pulls gcc and autotools, but not all builds require these dependencies. Some simple shell scripts and builds in other languages redundantly pull these dependencies as part of the build, even when builtins.derivation provided by nix language could be sufficient.
+
+The automatic header include and ldflag magic provided by stdenv often leads to a "just-works" mentality, causing us to become overly dependent on it.
+
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Everything is opt-in!
+
+Decoupling the system from fixed parts can be done by availing configuration options under the NixOS modules:
+
+```nix
+{
+  # bootloader.grub.enable = false;
+  bootloader.gummiboot.enable = true;
+  init.runit.enable = true;
+  boot.kernelPackages = pkgs.linuxPackages_latest;
+  boot.baseFiles = pkgs.baseFiles_latest;
+  # Other system configurations can be added here  ...
+}
+```
+
+This is an example constituting how one can easily replace grub with gummiboot.
+
+By providing this level of granularity, users can precisely control their system components, opting for their preferred choices while excluding unnecessary defaults. This empowers users to create tailored configurations and enhances the flexibility of their NixOS setups.
+
+### Emphasis on flake over shell script wrappers
+
+[nixos-rebuild](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh) is nothing but a complex shell script, that has to be present in the system beforehand, which basically calls `nix-build` under the hood, building a derivation, and then running the built derivation. All the options can be instead defined in nix language, in the flake definition itself, and the output derivation can be built using the nix command directly.
+
+[nixos-generators](https://github.com/nix-community/nixos-generators#using-in-a-flake) has quite recently started adding flake support through which there's no need to install or do anything of it, a simple `nix build` will build the iso/vm images. Making it transparent as well as reducing the extra script installation setup.
+
+### CLI improvements
+
+This is majorly for improving the binary cache user interaction, remote-builders setup, and cache serving.
+
+This is not the exact design required, but something like this will suffice:
+
+```bash
+nix-store -r /nix/store/{some.drv}        ## OR: nix build /nix/store/{some.drv}
+# pulls/builds whatever is in $(nix show-derivation /nix/store/{some.drv} | jq '.[] | .inputSrcs, .inputDrvs')
+# builds and produces output
+
+nix-store --export /nix/store/{somepath} > file
+
+## This is the main point, pull paths from binary cache, if we exactly know it
+nix store pull $(nix-store --query --requisites /nix/store/{some.drv})
```

Comment:
Can't this be done with `nix copy`?

Reply:
Wait, is that so? I didn't knew a binary cache can be used as `--from` on nix copy, all the examples in the documentation https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-copy.html seems to show `file://` or `ssh://` (machine-to-machine copy, not involving binary cache). Since this works, I think this is redundant to look for then...

Reply:
`nix copy` is for copying closure between two nix stores and cannot facilitate upstream caching, which substituter setup do; fallback to next substituter if path doesn't exist in one. So its not exactly the most generic solution (similar situation also mentioned in one of the [below comment](https://github.com/NixOS/rfcs/pull/157#discussion_r1281835622)).